docs/markdown/howtox.md

```diff
@@ -28,24 +28,36 @@ native-files and the latter via the cross file only.
 
 ## Set dynamic linker
 
+Like the compiler, the linker is selected via the <compiler variable>LD
+environment variable, or through the `<compiler entry>ld` entry in a native
+or cross file. You must be aware of whether you're using a compiler that
+invokes the linker itself (most compilers including GCC and Clang) or a
+linker that is invoked directly (when using MSVC or compilers that act like
+it, including Clang-Cl). With the former `cld` or `CCLD` should be the value
+to pass to the compiler's special argument (such as `-fuse-ld` with clang and
+gcc), with the latter it should be an exectuable, such as `lld-link.exe`.
+
+*NOTE* In meson 0.53.0 the `ld` entry in the cross/native file and the `LD`
+environment variable was used, this resulted in a large number of regressions
+and was changed.
+
 ```console
-$ CC=clang LD=lld meson <options>
+$ CC=clang CCLD=lld meson <options>
 ```
 
 or
 
 ```console
-$ CC=clang-cl LD=link meson <options>
+$ CC=clang-cl CCLD=link meson <options>
 ```
 
-Like the compiler, the linker is selected via the LD environment variable, or
-through the `ld` entry in a native or cross file. You must be aware of
-whehter you're using a compiler that invokes the linker itself (most
-compilers including GCC and Clang) or a linker that is invoked directly (when
-using MSVC or compilers that act like it, including Clang-Cl). With the
-former `ld` or `LD` should be the value to pass to the compiler's special
-argument (such as `-fuse-ld` with clang and gcc), with the latter it should
-be an exectuable, such as `lld-link.exe`.
+or in a cross or native file:
+
+```ini
+[binaries]
+c = 'clang'
+cld = 'lld'
```

Comment:
Is there discussion somewhere about the choice to use a `binaries` entry for something that isn't the name of a program?

To be clear: `lld` here happens to be a program name and a linker identifier. In many cases, though, this entry would not be a program name at all (for example, to pick `bfd` when using gcc).

It seems reasonable things that aren't binaries shouldn't be in the `binaries` section. Something like https://github.com/mesonbuild/meson/issues/6442#issuecomment-573302028 seems like a reasonable approach if we want to specify a program.

Reply:
This is an oddity of how meson is implemented. For some compiler/linker combinations we drive the linker through the compiler, others we don't. We invoke gnu ld and apple ld through gcc or clang, but we invoke link.exe and lld-link.exe directly. Having two different ways to do the same thing (pick which linker) seems really annoying and prone to "Why does this not work?" kind of issues.

Reply:
And also, "why do I put the compiler in binaries, but the linker in a different section?" Is a total valid argument. If we were doing this from scratch with what we know now, we would probably split the "binaries" section into four separate sections, a "compilers", a "linkers", a "binaries", and a "programs", where "binaries" is specifically for meson internal binaries pkg-config, cmake, llvm-config, etc, and "programs" is for things that `find_program` looks up.

Reply:
Regardless of section name, as suggested [here](https://github.com/mesonbuild/meson/issues/6442#issuecomment-573231412), could better names of new variables be used?
C and C++ compiler are specified in cross file by `c` and `cpp` variables, and their arguments by `c_args` and `cpp_args`, **not** `cargs` and `cppargs`, so names `cld` and `cppld` would be inconsistent. Please see [this](https://mesonbuild.com/Reference-tables.html#language-arguments-parameter-names).

Since `bfd` or `gold` are linker **identifiers**, and not exact filenames, I suggest to include `id` / `ID` in names of new variables.

Arguments passed to linkers are already specified in `c_link_args`, `cpp_link_args` etc., so something similar would make sense.

For `pkgconfig`, environment variable is `PKG_CONFIG`, so using underscore is not unprecedented.

Summary of new variables (in cross file and environment) suggested by me:
* `c_linker_id`: `C_LD_ID`
* `cpp_linker_id`: `CXX_LD_ID`
* `cs_linker_id`: `CS_LD_ID` (If it makes sense... A pair of variables for this language was omitted in original patch.)
* `d_linker_id`: `D_LD_ID`
* `fortran_linker_id`: `F_LD_ID`
* `objc_linker_id`: `OBJC_LD_ID`
* `objcpp_linker_id`: `OBJCXX_LD_ID`
* `rust_linker_id`: `RUST_LD_ID`